In [1]:
import pandas as pd

In [2]:
base = pd.read_csv("penguins.csv")
base

Unnamed: 0    species     island  bill_length_mm  bill_depth_mm  \
0             1     Adelie  Torgersen            39.1           18.7   
1             2     Adelie  Torgersen            39.5           17.4   
2             3     Adelie  Torgersen            40.3           18.0   
3             4     Adelie  Torgersen             NaN            NaN   
4             5     Adelie  Torgersen            36.7           19.3   
..          ...        ...        ...             ...            ...   
339         340  Chinstrap      Dream            55.8           19.8   
340         341  Chinstrap      Dream            43.5           18.1   
341         342  Chinstrap      Dream            49.6           18.2   
342         343  Chinstrap      Dream            50.8           19.0   
343         344  Chinstrap      Dream            50.2           18.7   

     flipper_length_mm  body_mass_g     sex  year  
0                181.0       3750.0    male  2007  
1                186.0       3800.0  female  2007  
2                195.0       3250.0  female  2007  
3                  NaN          NaN     NaN  2007  
4                193.0       3450.0  female  2007  
..                 ...          ...     ...   ...  
339              207.0       4000.0    male  2009  
340              202.0       3400.0  female  2009  
341              193.0       3775.0    male  2009  
342              210.0       4100.0    male  2009  
343              198.0       3775.0  female  2009  

[344 rows x 9 columns]

In [5]:
base.columns

Index(['Unnamed: 0', 'species', 'island', 'bill_length_mm', 'bill_depth_mm',
       'flipper_length_mm', 'body_mass_g', 'sex', 'year'],
      dtype='object')

In [6]:
base.shape

(344, 9)

Essa base apresenta dados de pinguins de três espécies. Vamos verificar quantos pinguins foram observados em cada espécie e qual a sua proporção em porcentagem?

Para isso, podemos montar uma pequena tabela de frequência. Primeiro precisa-se contar quantos pinguins, depois calcular a porcentagem com o parâmetro “normalize” definido como “True” para retornar a frequência relativa dividindo todos os valores pela soma dos valores, e por fim, montar a tabela (figura 5.2).

In [7]:
frequencia=base['species'].value_counts()

frequencia


Adelie       152
Gentoo       124
Chinstrap     68
Name: species, dtype: int64

normalize=True: Este parâmetro, quando definido como True, normaliza os resultados, ou seja, retorna as frequências relativas (proporções) em vez das contagens absolutas. Cada valor único na coluna 'species' terá uma porcentagem associada à sua ocorrência em relação ao total de observações.

In [8]:
porcentagem = base['species'].value_counts(normalize = True )*100

porcentagem

Adelie       44.186047
Gentoo       36.046512
Chinstrap    19.767442
Name: species, dtype: float64

In [9]:
distr_frequencia = pd.DataFrame({'Frequência': frequencia, 'Porcentagem (%)': porcentagem})

print(distr_frequencia)

           Frequência  Porcentagem (%)
Adelie            152        44.186047
Gentoo            124        36.046512
Chinstrap          68        19.767442


In [12]:
ILHA =base['island'].value_counts()

In [14]:
porcentagemilha = base['island'].value_counts(normalize = True )*100

In [15]:
ILHA

Biscoe       168
Dream        124
Torgersen     52
Name: island, dtype: int64

In [16]:
porcentagemilha

Biscoe       48.837209
Dream        36.046512
Torgersen    15.116279
Name: island, dtype: float64

In [19]:
distr_frequencia2 = pd.DataFrame({'Frequência': ILHA, 'Porcentagem (%)': porcentagemilha})

In [20]:
distr_frequencia2

Frequência  Porcentagem (%)
Biscoe            168        48.837209
Dream             124        36.046512
Torgersen          52        15.116279

In [21]:
base['body_mass_g'].count()


342

In [22]:
minimo = base['body_mass_g'].min()

minimo

2700.0

In [23]:
maximo = base['body_mass_g'].max()

maximo

6300.0

Calcular a média
A média aritmética, ou simplesmente média, é a medida de tendência central mais conhecida e utilizada para resumir a informação contida em um conjunto de dados (VIEIRA, 2017).

A média de um conjunto de dados é obtida somando todos os dados e dividindo o resultado pela quantidade.

Média é igual a Soma de todos os dados dividido pelo Total (quantidade). Ou média de x igual a somatória de x dividido por n.
A média é igual ao somatório de x, dividido por n.

Agora, vamos descobrir o peso médio dos pinguins e para isso precisamos indicar a coluna. Para facilitar utilizaremos a biblioteca Pandas.

In [24]:
media = base['body_mass_g'].mean()

media

4201.754385964912

Observe que a média foi calculada considerando 342 e não 344 pois o Pandas por padrão exclui os valores nulos ao calcular o resultado.

Também, podemos observar a média de pesos considerando o sexo dos pinguins utilizando por exemplo a função Groupby()

In [25]:
media = base.groupby(['sex'])['body_mass_g'].mean()

media

sex
female    3862.272727
male      4545.684524
Name: body_mass_g, dtype: float64

In [26]:
media = base.groupby(['species'])['body_mass_g'].mean()

media

species
Adelie       3700.662252
Chinstrap    3733.088235
Gentoo       5076.016260
Name: body_mass_g, dtype: float64

Descobrir a moda
Moda é o valor que ocorre com maior frequência. No caso da variável peso:

In [27]:
moda=base['body_mass_g'].mode()

moda

0    3800.0
Name: body_mass_g, dtype: float64

Assim, o peso 3800 gramas apareceu mais vezes e para saber quantas vezes basta verificar:

In [28]:
base['body_mass_g'].value_counts()

3800.0    12
3700.0    11
3900.0    10
3950.0    10
3550.0     9
          ..
3875.0     1
3275.0     1
4475.0     1
3975.0     1
3825.0     1
Name: body_mass_g, Length: 94, dtype: int64

Descobrir a mediana
A mediana divide um conjunto de dados ordenados em duas metades, ou seja, é o valor que ocupa a posição central do conjunto de dados ordenados (VIEIRA, 2017).

Para calcular a mediana dos pesos dos pinguins utilizaremos a função median() do Pandas

In [29]:
mediana = base['body_mass_g'].median()

mediana

4050.0

Isso quer dizer que o valor 4500 ocupa a posição central quando os valores dessa variável estão ordenados. Isso quer dizer que metade dos dados nesta variável estão acima ou abaixo desse valor e podemos comparar esse valor com a média calculada anteriormente observando que possivelmente há valores extremos que estão “puxando” a média.

Observar os quartis
Um conjunto de dados pode ser dividido em mais vezes, não somente em duas como no caso da mediana.

Os quartis dividem um conjunto de dados em quatro partes iguais. O segundo quartil é o mesmo que a mediana e o quarto quartil refere-se ao valor máximo.

Após encontrar os valores dos quartis pode-se calcular a distância interquartil: Terceiro quartil – Primeiro Quartil.

Continuando a observação da variável “body_mass_g” os quartis calculados foram:

Primeiro quartil: Para encontrar o primeiro quartil deve-se informar 25%, ou seja, 0.25.

In [30]:
quartis = base['body_mass_g'].quantile(q=0.25)

quartis

3550.0

Segundo quartil: Deve-se informar 50%, ou seja, 0.50 e assim, encontramos o mesmo valor ao calcular a mediana.

In [31]:
quartis = base['body_mass_g'].quantile(q=0.5)

quartis

4050.0

Terceiro quartil: Deve-se informar 75%, ou seja, 0.75

In [32]:
quartis = base['body_mass_g'].quantile(q=0.75)

quartis

4750.0

Quarto quartil: Informe 100%, ou seja, 1,00 e assim, encontramos o valor máximo que já tínhamos calculado anteriormente.

In [33]:
quartis = base['body_mass_g'].quantile(q=1.00)

quartis

6300.0

Intervalo interquartil
O intervalo interquartil é a diferença entre o terceiro e o primeiro quartis. Assim, considera somente os 50% dos dados e o resultado não é influenciado pelos valores discrepantes, pois os valores discrepantes estão localizados nos 25% de cada lado da distribuição.

Observando a variável “body_mass_g” a distância interquartil é:

Terceiro quartil – Primeiro Quartil:

In [34]:
4750 - 3550

1200

Desvio médio absoluto
O termo “desvio” trata-se da diferença absoluta entre um valor e a média dos dados, ou seja, estamos interessados na distância de um valor em relação à média.

Como devemos considerar o módulo (valor absoluto) dos desvios, evitando, deste modo, valores negativos de alguns desvios. Portanto, o desvio médio absoluto será́ dado por:

DM igual a somatória (sendo i igual a um), abre barra vertical - xi menos a média de x - fecha barra vertical, dividido por n.


A biblioteca Pandas fornece um método para tornar esse cálculo mais fácil, mad(), esse método exclui os valores ausentes ao calcular o resultado.

O desvio médio da variável “body_mass_g” é:

In [35]:
base.body_mass_g.mad()

<ipython-input-35-00c4b673d5e8>:1: FutureWarning: The 'mad' method is deprecated and will be removed in a future version. To compute the same result, you may do `(df - df.mean()).abs().mean()`.
  base.body_mass_g.mad()


676.3953011182928

Assim, a distância absoluta entre os valores e a média é aproximadamente 676,39. Podemos comparar a dispersão dos dados entre variáveis, considerando que quanto maior o valor do desvio maior a dispersão dos dados.

Um dos problemas de calcular o desvio médio é a dificuldade matemática que o módulo apresenta, claro a biblioteca do Python nos ajuda muito nesse quesito, mas há outra maneira de contornar esse pequeno problema no qual todos os valores são “elevados” ao quadrado assim, calculamos a média dos quadrados dos desvios que chamamos de variância.

Variância
A variância informa como os valores de uma variável se distribuem em torno da média. Assim, quanto maior a variância, mais longe da média estarão alguns valores.

In [36]:
base.body_mass_g.var()

643131.0773267478

Desvio Padrão
Geralmente os dados têm unidades de medidas como quilogramas, metros e entre outros e ao elevar os desvios ao quadrado essas unidades também são calculadas ao quadrado assim, dificultando a comparação do valor da variância com os dados. Para resolver isso, podemos extrair a raiz quadrada do resultado da variância que é chamado desvio padrão. Quando o desvio padrão está alto demais em relação às medidas de centralidade (média e mediana) sugere que há valores extremos incomuns que estão afetando a distribuição de valores em torno da média (AKANIME; YAMAMOTO, 2013).

O desvio padrão populacional é a raiz quadrada da variância populacional, ou seja:

Sigma igual a raiz de sigma ao quadrado.
E, o desvio padrão amostral (s) é a raiz quadrada da variância amostral, ou seja:

s igual a raiz de s ao quadrado.
A biblioteca Pandas, também, fornece um método que retorna o desvio padrão o std(), assim o desvio padrão calculada da variável “body_mass_g” é:

In [37]:
base.body_mass_g.std()

801.9545356980955

Coeficiente de Variação
Quando precisamos comparar duas distribuições, por exemplo, não podemos fazer uma comparação direta, uma vez que a magnitude dos dados dessas distribuições pode ser muito diferente. Podemos calcular o coeficiente de variação que é uma medida que expressa a variabilidade dos dados, mas exclui a influência da ordem de grandeza.

O coeficiente de variação analisa a dispersão em termos relativos em porcentagem (%). Quanto menor for o seu valor mais homogêneo serão os dados, ou seja, menor será a dispersão em torno da média. Para calcular divide-se o desvio padrão pela média:

cv igual a sigma dividido por u ou cv igual a s dividido por x.
Podemos calcular o coeficiente de variação da variável “body_mass_g” com o seguinte código:

In [38]:
coefVar = (base.body_mass_g.std())/(base['body_mass_g'].mean())*100

print(f' Coeficiente de variação: {coefVar:.2f} %')

 Coeficiente de variação: 19.09 %


Função describe()
A biblioteca Pandas fornece um método que retorna alguns detalhes estatísticos básicos de uma única vez (contagem, média, desvio padrão, mínimo, quartis e máximo): describe( ). A utilizaremos para a variável “body_mass_g”

In [39]:
base['body_mass_g'].describe()

count     342.000000
mean     4201.754386
std       801.954536
min      2700.000000
25%      3550.000000
50%      4050.000000
75%      4750.000000
max      6300.000000
Name: body_mass_g, dtype: float64